src/material_pool.hpp

```diff
@@ -8,160 +8,163 @@
-		void SetNormal(TextureHandle normal);
+		template<MaterialConstantType type>
+		typename MaterialConstantHelper<type>::Type GetConstant(){
+			return MaterialConstantHelper<type>::get_value(m_material_data.m_constant_data + (uint32_t(type) >> 16));
+		}
 
-		void UseNormalTexture(bool use_normal);
+		template<MaterialConstantType type>
+		void SetConstant(const typename MaterialConstantHelper<type>::Type& value) {
+			MaterialConstantHelper<type>::set_value(m_material_data.m_constant_data + (uint32_t(type) >> 16), value);
+		}
 
-		TextureHandle GetRoughness() { return m_rougness; }
-		void SetRoughness(TextureHandle roughness);
-
-		void UseRoughnessTexture(bool use_roughness);
-
-		TextureHandle GetMetallic() { return m_metallic; }
-		void SetMetallic(TextureHandle metallic);
-
-		void UseMetallicTexture(bool use_metallic);
+		TexturePool* const GetTexturePool() { return m_texture_pool; }
 
-		TextureHandle GetAmbientOcclusion() { return m_ao; }
-		void SetAmbientOcclusion(TextureHandle ao);
+		ConstantBufferHandle* const GetConstantBufferHandle() const { return m_constant_buffer_handle; };
+		void SetConstantBufferHandle(ConstantBufferHandle* handle) { m_constant_buffer_handle = handle; };
 
-		void UseAOTexture(bool use_ao);
+		void UpdateConstantBuffer();
 
-		DirectX::XMFLOAT3 GetConstantAlbedo()
+		union TextureFlags
 		{
-			return
-			{
-				DirectX::XMVectorGetX(m_material_data.m_color),
-				DirectX::XMVectorGetY(m_material_data.m_color),
-				DirectX::XMVectorGetZ(m_material_data.m_color)
+			struct {
+				uint32_t m_has_albedo_texture : 1;
+				uint32_t m_has_normal_texture : 1;
+				uint32_t m_has_roughness_texture : 1;
+				uint32_t m_has_metallic_texture : 1;
+				uint32_t m_has_ao_texture : 1;
 			};
-		};
-
-		void SetConstantAlbedo(DirectX::XMFLOAT3 color);
 
-		bool UsesConstantAlbedo() { return m_material_data.m_material_flags.m_use_albedo_constant; };
-		void SetUseConstantAlbedo(bool use_constant_albedo) { m_material_data.m_material_flags.m_use_albedo_constant = use_constant_albedo; };
-
-		float GetConstantAlpha() { return DirectX::XMVectorGetW(m_material_data.m_color); };
-		void SetConstantAlpha(float alpha);
+			uint32_t m_value;
+		};
 
-		DirectX::XMFLOAT3 GetConstantMetallic()
+		union MaterialData
 		{
-			return
-			{
-				DirectX::XMVectorGetX(m_material_data.m_metallic_roughness),
-				DirectX::XMVectorGetY(m_material_data.m_metallic_roughness),
-				DirectX::XMVectorGetZ(m_material_data.m_metallic_roughness)
-			};
-		};
+			struct {
 
-		void SetConstantMetallic(DirectX::XMFLOAT3 metallic);
+				float m_color[3];
+				float m_metallic;
 
-		bool UsesConstantMetallic() { return m_material_data.m_material_flags.m_use_metallic_constant; };
-		void SetUseConstantMetallic(bool use_constant_metallic) { m_material_data.m_material_flags.m_use_metallic_constant = use_constant_metallic; };
+				float m_roughness;
+				float m_is_double_sided;
+				float m_use_alpha_constant;
```

Comment:
same as before, this is alpha masked not constant, if you want to support both then we need two flags